[RFC 0189] Contracts (#189)
NixOS/rfcs: rfcs/0189-contracts.md

Some examples of possible contracts:

- File backup
- Streaming backup (for databases)
- Secrets (out of store values) provisioning
- SSL certificate generation
- Database setup (ensuring a database and user exist)
- Reverse proxy setup
- Reverse proxy "chain" allowing transparent traffic inspection
- LDAP user and group management
- OIDC provider integration
- Forward auth setup

Any implicit convention in nixpkgs can be encoded this way.

This RFC's goal is _not_ to define all these contracts
nor to identify an exhaustive list of existing contracts, but to define a pattern derived from a few diverse examples.

These contracts will live under a new option path `contracts`
like `contracts.fileBackup` and `contracts.streamingBackup`.

See [prior art] for some useful comparisons that can help you get a better picture.

# Implementation

The implementation was worked out initially in the [SelfHostBlocks] repo and perfected in the [module interfaces] repo.
There are some slight variations proposed in this RFC relative to the module interfaces repo to get it out sooner rather than later. See the [corresponding unresolved section](#dual-link).

> It is important to keep in mind that the proposed implementation comes from

Reviewer: To me, this proposal is screaming "it is a convention to do dependency injection consistently for modules".
"Dependency injection" is a mean to achieve dependency inversion that could (should?) be very familiar to developpers.

It is mentionned in the [Prior Art](https://github.com/ibizaman/rfcs/blob/contracts/rfcs/0189-contracts.md#prior-art) section, but I feel that it could be mentionned here, and even appear earlier (as early as in [Motivation](https://github.com/ibizaman/rfcs/blob/contracts/rfcs/0189-contracts.md#motivation)).

Author: I didn’t see it at first but it’s evident to my too now

where the resulting file is created with the requested `owner`, `mode`, etc.
as described by the contract `consumer`.

This new provider has been tested using the [contract's behavior test](https://github.com/NixOS/nixpkgs/pull/432529/commits/448410a520225bc71e1616611cef7ad086c64cd1)
and has been used in [`services.stash`'s module](https://github.com/NixOS/nixpkgs/pull/432529/commits/19419ad95913fbed4636d0b24d95c80517c18340) as an example.

# Drawbacks

We are not aware of any because this solution is fully backwards compatible,
incremental, and has many advantages. It also arose from a real practical need.

Care should be taken to not abuse this pattern though. It should be reserved
for contracts where abstracting away a `consumer` and `provider` makes sense.
We didn't find a general rule for that, but a good indicator of an unnecessary contract is where we only find one instance of a `consumer` and `provider` pair in the whole of nixpkgs.

# Alternatives

This design arose from trying to maximize code reuse.
We started by fiddling with nix code and the implementation emerged naturally.

We are not aware of any alternative ways to do this,

> mostly because our attempts to tweak the code often led us often to infinite recursion or other module issues

Reviewer 2: What about doing something along the lines of this:
```nix
config = {
  
  # each service coming from submodule generator units.xyz is where the magic happens, its fully isolated. I would expect it to run its own modules evaluator in the background, should even help with eval time
  services.nextcloud = units.nextcloud {
     # ... module config ...
     fileBackup.provider = config.services.restic-for-nextcloud.fileBackup;
  };
   
  # Service name does not matter
  services.restic-for-nextcloud = units.restic {
    // Provider-specific options.
    repository = "/var/lib/backups/nextcloud";
    passwordFile = toString (pkgs.writeText "password" "password");
    initialize = true;
  };
};
```

This way we make modules more portable, contract itself is with root module system and units.xyz which we could implement to support other projects such as home-manager or devshells. and integration between modules is dependent on what other modules explicitly export and ofcourse where user explicitly  passes values

Tho there might be cases of infinite recursion sometimes, tho it should be very rare

I think of this as Terraform modules, they have certain inputs and certain outputs - everything else is a black box

Reviewer 3: this also reminds me a bit of [portable services](https://github.com/NixOS/nixpkgs/pull/267111), and reusing those for the existing modules to prevent duplication.
this example does sound easily extendible in terms of adding new providers - @ibizaman would probably know more about how we were doing there so far.
avoiding dual linkage maybe seems an open question, but yeah, let's try stuff.

Author: Same, makes me think of portable services.

I fully agree we should tend to this and it seems indeed compatible with contracts. I would argue though it is orthogonal and a large effort that should be separate from this.

I'm definitely not going to stop anyone from attempting both at the same time but I can't see myself doing it currently.

- 04/2024: Scale21x in Pasadena: [Easier NixOS self-hosting with module contracts](https://www.youtube.com/watch?v=lw7PgphB9qM)
- 11/2024 at NixCon2024 in Berlin: [Enabling incremental adoption of NixOS with module contracts](https://www.youtube.com/watch?v=CP0hR6w1csc)

A pre-RFC has been opened [on discourse](https://discourse.nixos.org/t/pre-rfc-decouple-services-using-structured-typing/58257).

A few useful comparisons beyond nixpkgs:

- Contracts are closely related to Golang interfaces with options being methods and input and output options the inputs and outputs of the methods.
  The important bit is that in Golang, the saying goes "the bigger the interface, the weaker the abstraction".
  We should strive to keep the number of options to a minimum to make the contracts more general.
- Contracts are reminiscent of the [reverse dependency principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle) which is used in many places.

# Unresolved questions

## Dual Link

The current implementation requires the `end user` to link the consumer and provider
in both directions:

```nix

> config = {

Reviewer 4: In my understanding consumer and provider are switched here.

- nextcloud don't need a fileBackup, it provides a fileBackup (I assume these are files that should be backed up). => nextcloud is a provider
- but restic (and borgBackup) need to know which files to backup (fileBackup) so they would consume these files. => restic is a consumer

I would also argue:  a provider provides only one thing but a consumer can consume multiple things.

Maybe this is what is meant here, and I mix up the syntax (or the object naming could be updated), but I couldn't see descriptions on your option definition up there.

Reviewer 4: First of all sorry, if this was already discussed. I read the last hour the discusions specifically on this topic, but couldn't find anything. But I might have missed it.

Anyway I'm a huge dependency injection fan. But I don't see to much benefit on your example here.
Usually dependency injection comes with a injector which handles the wiring for you, which is one of the benefits.
But we have to make the wireing here our self.
The nixos module system already comes with options which can be rewired on demand where ever you like
(outside the module definitions),
which is not possible in programming languages like java or go.
For example your example currently in nixos could be done like this

```nix
services.restic.bacups.nextcloud.paths = [config.services.nextcloud.datadir]
```

This approach don't lack any aspect to using the contracts (At least from my perspective).

The one benefit I see from the contracts concept is that we create a transactional type system, which standardizes service plumbing. Which is not nothing.

If you would tell me nixos module is capable of dependency injection, I would imaging something like

```nix
options.services.nextcloud.datadir = mkProvider fileBackup {
# same like mkOption but type is dictated by fileBackup (e.g.: 2 fileBackup options can't have different types)
};
...
config = .. # nextcloud configuration using datadir like now
```

than you could set.

```nix
serivces.restic.backup.dependencyInjection.paths = allProvidersOf fileBackup;
```

In this example `allProvidersOf <key/type>` will be handled by the dependency injector and
creates a list of config-references.
```nix
services.restic.backup.dependencyInjection.paths = [config.services.nextcloud.datadir];
```
This is because I expect automatic wireing when I hear dependency injection = inversion of control.

I sketched some scenarios (e.g.: reverse proxy -> webapp -> database connetion <- credentials) and I come always to the same conclusion. But I could see a benefit in the credentials part with this approach, but this is not part of you RFC if I understood correctly.

I hope this is helpful.

p.s.: I would be happy for a proper solution here, I faced a similar problem like in your example and solved it like this : https://gist.github.com/mrVanDalo/a1f7ec5cce80d69ac60b6d22feb6d768 (a home manager interface which handles files I need to backup)

Reviewer 3: thanks for your response!

i think the dual link here was maybe less intended design, more like a limitation of what we'd managed with the implementation at the time. #506343 addresses that.

> `mkProvider fileBackup { ... }`

the limitation of functions is that they are pure, whereas going thru the module system ensures that we can use the passed information to set `config` as well.

> In this example `allProvidersOf <key/type>` will be handled by the dependency injector and creates a list of config-references.

this type of aggregation is basically what #506343 does, resulting in the `defaultProvider(Name)` to let one wire things up.

on naming, i def had trouble picking things that made sense as well. in the case of secrets, for one, i sort of made sense out of this by speaking of an application _requesting_ a secret, that another module might then _provide_.
in an earlier iteration, i played with terms like input and output, but then ran into similar issues about perspective as you noted here.

Reviewer 3: @mrVanDalo:

> `mkProvider fileBackup { ... }`

at NixOS/nixpkgs#506343 i've now managed to cut down on boilerplate further again using a `mkProviderType` helper now, perhaps not unlike what you'd aimed for with your mentioned `mkProvider`.

> I sketched some scenarios (e.g.: reverse proxy -> webapp -> database connetion <- credentials)

fwiw, i added a test (`nixos/tests/contracts/chaining.nix`) to my commit `nixos/modules/contracts: add NixOS contracts integration` that demonstrates chaining of contracts now.

> `allProvidersOf`

i put up a [branch](https://github.com/KiaraGrouwstra/nixpkgs/compare/contracts-automated...contracts-all-results) exploring this, tho i'm a bit on the fence on when one might want to use this still

> I faced a similar problem like in your example and solved it like this : https://gist.github.com/mrVanDalo/a1f7ec5cce80d69ac60b6d22feb6d768 (a home manager interface which handles files I need to backup)

i think this does sound a bit similar to how i aggregate things in my PR. now, back-ups i think don't make full use of the concept of contracts in the sense i think they don't really need to get a `result` value back in nix, which the contracts facilitate. this is also how we got the idea of splitting the back-ups out of an initial contracts PR.

Reviewer 4: `nixos/tests/contracts/chaining.nix` is quite helpful to understand your setup. 
I added some questions, which might help me  (and others) to understand better what's happening.

## Data Flow
[dataflow]: #dataflow

Another consideration before looking at the code is how data flows through a contract.

```mermaid
sequenceDiagram
  participant Consumer as Consumer
  participant Contract as Contract
  participant Provider as Provider
  participant EndUser as End User
  participant Config as Config
  autonumber
  Consumer ->> Contract: set input
  Contract ->> Provider: read input
  opt
    EndUser ->> Provider: set option
  end
  Provider ->> Config: do side effect
  opt
    Provider ->> Contract: set ouput
  end
  opt
    Contract ->> Consumer: read output
  end
```

1. A `consumer` sets the `input` option of the contract.
2. The `provider` reads from that `input` option.
3. The `provider` optionally accepts provider-specific options set by the `end user`.
4. The `provider` does some side effect (otherwise, there's no point).
5. The `provider` optionally writes to the `output` of the contract.
6. The `consumer` optionally reads from the `output` of the contract.

If you squint, this looks just like function application, only applied at the module level.

## Contract Interface
[Contract Interface]: #contract-interface

> _The draft PR from which the following snippets are taken can be found [here][draft]._

Reviewer 3: the link here doesn't render this way

[agenix]: https://github.com/ryantm/agenix
[sops-nix]: https://github.com/Mic92/sops-nix

Contrary to the previous contracts we covered, the `consumer` here needs to read the `output` of the `provider`
because it contains the path to the file containing the secret.

When testing a module that expects a file containing a raw secret,
the ubiquitous method to provide the file is using `pkgs.writeText`.
This works but has the issue the created file is world readable
and we thus do not test the file is accessible with the correct user.
To avoid this pitfall going forwards, we created the `testing.hardcodedSecret`
`provider` which is an improved version of `pkgs.writeText`
where the resulting file is created with the requested `owner`, `mode`, etc.
as described by the contract `consumer`.

This new provider has been tested using the contract' behavior test
and has been used in `services.stash`'s as an example.

# Drawbacks
[drawbacks]: #drawbacks

We are not aware of any because this solution is fully backwards compatible,

> incremental and has a lot advantages. It also arise from a real practical need.

Reviewer 3: arises/arose

Care should be taken to not abuse this pattern though. It should be reserved
for contracts where abstracting away a `consumer` and `provider` makes sense.
We didn't find a general rule for that but a good indication that the pattern gets abused
is if we only find one `consumer` and `provider` pair in the whole of nixpkgs.

# Alternatives
[alternatives]: #alternatives

> This design arise from trying to minimize code reuse.

Reviewer 3: arises/arose, maximize code reuse

We started by fiddling with nix code and the implementation came up naturally.

We are not aware of any alternatives to do this,

> mostly because our attempts to tweak the code led us often to infinite recursion or other module issues

Reviewer 3: often led us

so we couldn't stray too far from the way it is written now.

# Prior art
[prior-art]: #prior-art

We did not find any discussion about any of this by the nix community.

> It is a bit self-centered but the two talks I (ibizaman) gave on this subject in nixpkgs can be considered prior arts.

Reviewer 3: (prior art  :no_mouth:)

> Note the syntax presented is a bit outdated now but the message and the passion is still relevant:

Reviewer 3: (it wasn't clear to me what you meant by 'passion' here)

Author: I mostly meant that passion I had behind this endeavor. :P But I agree it's superfluous.

- 04/2024: Scale21x in Pasadena: [Easier NixOS self-hosting with module contracts](https://www.youtube.com/watch?v=lw7PgphB9qM)
- 11/2024 at NixCon2024 in Berlin: [Enabling incremental adoption of NixOS with module contracts](https://www.youtube.com/watch?v=CP0hR6w1csc)

> A pre-RFC has been opened [on discourse][prerfc].

Reviewer 3: link failed to render

> [prerf]: https://discourse.nixos.org/t/pre-rfc-decouple-services-using-structured-typing/58257

Reviewer 3: prerfc

## Dual Link
[unresolved-duallink]: #unresolved-questions-duallink

A few useful comparisons outside of nixpkgs are:

- Contracts are closely related to Golang interfaces with options being methods and input and output options the inputs and outputs of the methods.
  The important bit is that in Golang, the saying goes "the bigger the interface, the weaker the abstraction".
  We should strive to keep the number of options to a minimum to make the contracts more general.
- Contracts are reminiscent of the [reverse dependency principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle) which is used in a lot of places.

# Unresolved questions
[unresolved]: #unresolved-questions

## Dual Link
[unresolved-duallink]: #unresolved-questions-duallink

The current implementation requires the `end user` to link the consumer and provider
in both directions:

```nix
config = {

>   // consumer to provider

Reviewer 3: `//` -> `#` (down as well)

Author: :man_facepalming:

Unfortunately, this has two downsides:

1. It requires one more line in each provider definition. This would be okay if there wasn't the following downside.
2. There's no way to write side effects. This means the `provider` can only write to its own `output`, which misses the whole point of having contracts in the first place.

There's maybe a way to solve this but we didn't figure it out. Help is appreciated!
Beware though you will be crossing the edge of the module system and entering the land of infinite recursion.

## Documentation
[unresolved-documentation]: #unresolved-questions-documentation

It is not possible to build the manual right now. Doing so results in an error.

```bash
$ (cd nixos/; nix-build release.nix -A manual.x86_64-linux)

[...]

       error: attribute 'contracts' missing
       at /home/timi/Projects/nixpkgs/nixos/modules/services/web-apps/stash.nix:435:16:
          434|       jwtSecretKeyFile = mkOption {
          435|         type = config.contracts.secret.consumer;
             |                ^
          436|         description = "Path to file containing a secret used to sign JWT tokens.";
```

> Comments in the draft PR have been added to indicate what has been tried.

Reviewer 3: link draft PR here

Reviewer 3: https://github.com/NixOS/nixpkgs/pull/432529

a lot of code and must be thought out
thoroughly to get it right and to test correctly.

Having this code live in each module separately
is a waste for the whole community.
We see at least those disadvantages:

- It's more code to review and maintain for everybody.
- More burden on maintainers of a module implementing this feature:
  they must know how to setup their dependencies at a low-level
  and must keep the code up to date.
- Leads to difference in interface:
  options to setup a same dependency are different across modules.
- Leads to difference in implementation:
  every maintainer has their own style and knowledge,
  leading to not every implementation being of the same quality
  and being tested equally.
- Leads to difference in features:
  some implementations are more featureful than others,
  very few modules allow you to choose from multiple dependencies (e.g. PostgreSQL, MySQL or other).
- Not extendable without changing the source code:
  a user cannot easily choose to use a dependency the maintainer didn't add code for.

What we propose answers to all those issues
as well as allows a few things that's not possible currently:

> - interfacing with dependencies and services outside of NixOS,

Reviewer 5: Is there an example for this? The contract system will also probably need integration with service managers for scheduling and bringing up dependencies.

Author: I can’t believe I forgot to add an example for this IMO important point. I’ll add that.

The intent here is to provide an escape hatch to the outside (of NixOS) world. Which means you’re free to do whatever you want. The use case I had in mind does imply “manual” scheduling. Let’s say you want to integrate a NixOS service with a database that exists already on another server which does not run NixOS. You would create a database provider which hardcodes the url and credentials to the other server. Then the consumer would be able to access it. Of course you must set up the other server beforehand correctly. And of course the other server could be a NixOS machine too.

I can see a separate contract for interacting with service management. This should be merged with the module interfaces initiative.

Author: A good example of this would be for a database contract. The options could be:

Filled out by the provider:
- `hostname`
- `database name`
- `username`
- `password`

I actually can't think of options filled out by the consumer. Maybe extensions needed in the database? Charset?

Usually, you would have a provider like the postgresql module that you would create a database in with the (unfortunately but understandably controversial) `ensureUser` and `ensureDatabase` options. Then, you would plug in the new user and database to the consumer that requests a database.

But this would work too outside of the NixOS ecosystem. You could just create a user and database in any system then, inside the NixOS system that has the consumer, just hardcode the 4 options with strings that would point to the other system.

Does that make sense?

Reviewer 5: I believe those fields are enough for most configurations, though the password should be specified using `passwordFile` (or a secret management contract!).

Author: That’s spot on @axelkar. I should’ve been more precise. In my project I indeed use a contract for the password itself so it’s not the plain password.

and we thus do not test the file is accessible with the correct user.
To avoid this pitfall going forwards, we created the [`testing.hardcodedSecret`
`provider`][hardcodedSecret: new secret contract consumer]
which is an improved version of `pkgs.writeText`
where the resulting file is created with the requested `owner`, `mode`, etc.
as described by the contract `consumer`.

[hardcodedSecret: new secret contract consumer]: https://github.com/NixOS/nixpkgs/pull/432529/commits/6fbd099aa306d2cce337b8fa7ed7e0c8a255aebf

This new provider has been tested using the [contract's behavior test][hardcodedSecret: define behavior test for secret contract]
and has been used in [`services.stash`'s module][stash: use secret contract for passwords] as an example.

[hardcodedSecret: define behavior test for secret contract]: https://github.com/NixOS/nixpkgs/pull/432529/commits/448410a520225bc71e1616611cef7ad086c64cd1
[stash: use secret contract for passwords]: https://github.com/NixOS/nixpkgs/pull/432529/commits/19419ad95913fbed4636d0b24d95c80517c18340

# Drawbacks
[drawbacks]: #drawbacks

We are not aware of any because this solution is fully backwards compatible,

> incremental and has a lot advantages. It also arose from a real practical need.

Reviewer 6: Standard question for any module-system-based increase in layers: what is the estimated evaluation-time resource impact?

Author: That’s a great question. Is there a standard set of benchmarking tooling? After some quick online search, I see some external repos and wiki guides but I was wondering how from scratch I should start from.

Reviewer 6: Good question; I don't actually know things beyond `export NIX_SHOW_STATS=1`. There are some scripts in `ci/eval` but I am not sure whether reusing parts of them is worth it (but maybe looking at what is done there could be useful)

Reviewer 3: maybe one way to compare would be getting such stats after/before the PR for say that nextcloud test, like: `NIX_SHOW_STATS=1 nix-build --impure nixos/tests/nextcloud -A basic31`.
  
## PR

```json
{
"cpuTime": 0.1535560041666031,
"envs": {
  "bytes": 2400992,
  "elements": 167207,
  "number": 132917
},
"gc": {
  "heapSize": 402915328,
  "totalBytes": 68045680
},
"list": {
  "bytes": 288200,
  "concats": 1487,
  "elements": 36025
},
"nrAvoided": 133723,
"nrFunctionCalls": 122934,
"nrLookups": 66643,
"nrOpUpdateValuesCopied": 1889724,
"nrOpUpdates": 6747,
"nrPrimOpCalls": 60323,
"nrThunks": 171874,
"sets": {
  "bytes": 35799712,
  "elements": 2221018,
  "number": 16464
},
"symbols": {
  "bytes": 335942,
  "number": 34889
},
"values": {
  "bytes": 18234936,
  "number": 759789
}
}
```
  
## parent revision

```json
{
"cpuTime": 0.1342640072107315,
"envs": {
  "bytes": 2304512,
  "elements": 160244,
  "number": 127820
},
"gc": {
  "heapSize": 402915328,
  "totalBytes": 67791424
},
"list": {
  "bytes": 277016,
  "concats": 1378,
  "elements": 34627
},
"nrAvoided": 128713,
"nrFunctionCalls": 118695,
"nrLookups": 63610,
"nrOpUpdateValuesCopied": 1915086,
"nrOpUpdates": 6556,
"nrPrimOpCalls": 58429,
"nrThunks": 161589,
"sets": {
  "bytes": 36067568,
  "elements": 2239118,
  "number": 15105
},
"symbols": {
  "bytes": 299281,
  "number": 30518
},
"values": {
  "bytes": 17928864,
  "number": 747036
}
}
```

i'm not sure which metrics are interesting (note `cpuTime` is unstable across invocations) or if we should compare for a different build attribute.
building the nextcloud module (`nixos/modules/services/web-apps/nextcloud.nix`) instead, for what it's worth, i didn't really get it to show differences beside on a `symbols` metric.

<!--   
## comparison
  
the `cpuTime` metric is unstable across invocations, so for these two attributes any difference appears negligible to the variance.

- `cpuTime`: %
- `envs.bytes`: 104.186569651189%
- `envs.elements`: 104.345248496044%
- `envs.number`: %
- `gc.heapSize`: %
- `gc.totalBytes`: %
- `list.bytes`: %
- `list.concats`: %
- `list.elements`: %
- `nrAvoided`: %
- `nrFunctionCalls`: %
- `nrLookups`: %
- `nrOpUpdateValuesCopied`: %
- `nrOpUpdates`: %
- `nrPrimOpCalls`: %
- `nrThunks`: %
- `sets.bytes`: %
- `sets.elements`: %
- `sets.number`: %
- `sizes.Attr`: %
- `sizes.Bindings`: %
- `sizes.Env`: %
- `sizes.Value`: %
- `symbols.bytes`: %
- `symbols.numbers`: %
- `values.bytes`: %
- `values.numbers`: %
-->

Care should be taken to not abuse this pattern though. It should be reserved
for contracts where abstracting away a `consumer` and `provider` makes sense.
We didn't find a general rule for that but a good indication that the pattern gets abused

> is if we only find one `consumer` and `provider` pair in the whole of nixpkgs.

Reviewer 5: This is confusing. Is this what you meant?
```suggestion
is if we only find one `consumer` and one `provider` usage of a contract in Nixpkgs.
```

Author: Yes indeed thats what I meant :)

# Alternatives
[alternatives]: #alternatives

This design arose from trying to maximize code reuse.
We started by fiddling with nix code and the implementation came up naturally.

We are not aware of any alternatives to do this,

> mostly because our attempts to tweak the code often led us often to infinite recursion or other module issues

Reviewer 5: The dual-linking approach strikes dangerously close to infinite recursion in my mind. What do you think?

Reviewer 3: in my understanding the reason the current implementation works is lazy evaluation

Author: Indeed, it's pretty close. It's working because we're linking two different options, depending on the direction.

I'll argue I'm not the first to have though of this pattern. Let's take the secrets contracts as an example. Remember, it looks like so:

```nix
services.myservice.secret.provider = config.sops.secrets."myservice/secret".provider;

sops.secrets."myservice/secret".consumer = services.myservice.secret.consumer;
```

If we compare to the sops-nix module, if we squint we can see the dual linking is nearly there:

```nix
services.myservice.secret = config.sops.secrets."myservice/secret".path;

sops.secrets."myservice/secret" = {};
```

Btw, you would get the infinite recursion if you removed the `.consumer` or `.provider` from the snippet above.

Reviewer 3: i mean, it works until you try and serialize the whole structure i guess :sweat_smile:

# Unresolved questions
[unresolved]: #unresolved-questions

## Dual Link
[unresolved-duallink]: #unresolved-questions-duallink

The current implementation requires the `end user` to link the consumer and provider
in both directions:

```nix
config = {
  # consumer to provider
  services.nextcloud.fileBackup.provider = config.services.restic.backups.nextcloud.fileBackup;

  # provider to consumer
  services.restic.backups.nextcloud.fileBackup.consumer = config.services.nextcloud.fileBackup;
};
```

It would be so much nicer if we could somehow require only to give the `consumer` to the `provider`
and it managed to make the link backwards automatically.
In the snippet above, this means removing the need for the `provider to consumer` line.

The issue comes from the `consumer` and `provider` option in the top-level `contracts` definition to be of type `optionType`.
They don't have access to the actual `input` and `output` values of an instantiated contract.

Experimenting on this has been done in the [module interfaces][] repo.

> There, we set the `provider` option as a function which takes an argument

Reviewer 3: maybe for context we could link 'set the `provider` option' to the [lines in question](https://github.com/fricklerhandwerk/module-interfaces/blob/caeb171fde3a6ce2dc61f4e59834df9e7bf584a0/interface.nix#L47-L58)

Beware though you will be crossing the edge of the module system and entering the land of infinite recursion.

## Documentation
[unresolved-documentation]: #unresolved-questions-documentation

It is not possible to build the manual right now. Doing so results in an error.

```bash
$ (cd nixos/; nix-build release.nix -A manual.x86_64-linux)

[...]

       error: attribute 'contracts' missing
       at /home/timi/Projects/nixpkgs/nixos/modules/services/web-apps/stash.nix:435:16:
          434|       jwtSecretKeyFile = mkOption {
          435|         type = config.contracts.secret.consumer;
             |                ^
          436|         description = "Path to file containing a secret used to sign JWT tokens.";
```

Comments in the [draft PR][draftPR] have been added to indicate what has been tried.
Any help is appreciated to solve this.

# Future work
[future]: #future-work

- Solve the [documentation][unresolved-documentation] issue.
- Identify contracts and their inputs, outputs and behavior tests.
- Identify services that would benefit from being consumers and providers of contracts and add the necessary options.

> - Optionally solve the [dual-link][unresolved-duallink] issue.

Reviewer 3: clicking these links seems not to actually navigate the user to these on the page - i think it wants the ids instead?

Author: Just checked it is fixed now :)
